File: src/model/osi_layers.rs
//! [`OSI`] layers
//!
//! This module contains the definition of networking activities and their properties.
//! We assume that these data structures are prone to change, given that these are configuration artifacts.
//!
//!  [`OSI`]: https://en.wikipedia.org/wiki/OSI_model

use serde::{Serialize, Deserialize};

use crate::model::status::DeviceStatus;


#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Port {
    serial_number: String,
    name: String,
    model: String,
    number: i64,
    mac: String,
    status: DeviceStatus,
    speed_gb: i64,
}

impl Port {
    pub fn is_up(&self) -> bool {
        self.status.is_up()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Link {
    left: Port,
    right: Port,
}

impl Link {
    pub fn is_up(&self) -> bool {
        self.left.is_up() && self.right.is_up()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum LinkLayer {
    Layer2,
    Layer3,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Protocol {
    TCP,
    UDP,
    ICMP,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct IpAddress {
    address: String,
    netmask: String,
    gateway: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Route {
    destination: IpAddress,
    gateway: IpAddress,
    metric: i64,
}


#[derive(Debug, Serialize, Deserialize)]
pub struct FirewallRule {
    name: String,
    action: String,
    source: IpAddress,
    destination: IpAddress,
    protocol: Protocol,
    port: i64,
}

File: horao/conceptual/osi_layers.py
# -*- coding: utf-8 -*-#
"""OSI layers

This module contains the definition of networking activities and their properties.
We assume that these data structures are prone to change, given that these are configuration artifacts.
"""
from __future__ import annotations

from enum import Enum, auto


class LinkLayer(Enum):
    Layer2 = auto()
    Layer3 = auto()


class Protocol(Enum):
    TCP = auto()
    UDP = auto()
    ICMP = auto()


class IpAddress:
    def __init__(self, address: str, netmask: str, gateway: str):
        self.address = address
        self.netmask = netmask
        self.gateway = gateway


class Route:
    def __init__(self, destination: IpAddress, gateway: IpAddress, metric: int):
        self.destination = destination
        self.gateway = gateway
        self.metric = metric


class FirewallRule:
    def __init__(
        self,
        name: str,
        action: str,
        source: IpAddress,
        destination: IpAddress,
        protocol: Protocol,
        port: int,
    ):
        self.name = name
        self.action = action
        self.source = source
        self.destination = destination
        self.protocol = protocol
        self.port = port
